### PR TITLE
sql: allow creation of tables with no primary key

### LINKED ISSUE #2081: sql: support tables without a primary key?
This could probably be done without too much effort by creating a hidden column that is given a uuid. I'm not sure why we'd want this in practice, though part of the sqllogictest suite uses tables with primary keys which might be reason enough.

### PR DESCRIPTION
**Pull request by madelynnblue**
fixes #2081

[Review on Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/3286)

### CODE REVIEW

**Comment by tamird** [comment]
Can you file the other bugs you found? I'd really like for this to go in with the associated logic test enablement.

File: sql/create.go [hunk 1]
```diff
@@ -126,7 +126,28 @@ func (p *planner) CreateTable(n *parser.CreateTable) (planNode, error) {
 	// Inherit permissions from the database descriptor.
 	desc.Privileges = dbDesc.GetPrivileges()
 
-	if err := desc.AllocateIDs(); err != nil {
+	err = desc.AllocateIDs()
+	if err == errMissingPrimaryKey {
```
Comment (tamird) [inline sql/create.go]: instead of relying on this error, can you check for the presence of a primary key directly?

**Comment by mjibson** [comment]
Sure.

File: sql/create.go (hunk 1 above)
Comment (mjibson) [inline sql/create.go]: I tried that, and it sadly doesn't work. I duplicated the check in `Validate` for no primary keys, but the fields it relies on to do that are populated near the end of `AllocateIDs`, right before it calls Validate itself. Hence the double call to AllocateIDs in this function.

Comment (tamird) [inline sql/create.go]: I don't follow; you can't do `if len(desc.PrimaryIndex.ColumnNames) == 0`? The check in `Validate` is checking that the IDs were assigned, but you can just look at the names here.

**Comment by petermattis** [comment]
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[sql/create.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z09bhdgQ-U0oH2jhv:-K4Z09bhdgQ-U0oH2jhw:1654146534) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L132)):*
Is this the right builtin to be using?  `experimental_unique_int()` would be faster, though the ordering might not be desirable.

---

*[sql/create.go, line 134 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z14FVNeZMMqbkETO1:-K4Z14FVNeZMMqbkETO2:1277412493) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L134)):*
Is there precedent for the leading underscore?  Perhaps just `ROWID` for this column name. Btw, column names are case-insensitive and we generally use lowercase.

---

*[sql/testdata/no_primary_key, line 10 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z22xbr8ikoHdTw-Ml:-K4Z22xbr8ikoHdTw-Mm:-793423954) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/testdata/no_primary_key#L10)):*
If you added a new column to the table via an `ALTER TABLE t ADD c INT` statement, then the internal columns would be `a`, `b`, `_rowid`, `c`. Subsequent insertions would not work as expected. I think you need to tag the `_rowid` ColumnDescriptor as auto-generated so that it can be ignored for insertions and other operations. 

---

*[sql/testdata/no_primary_key, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z1r3RhUy45_gL0NMm:-K4Z1r3RhUy45_gL0NMn:55614947) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/testdata/no_primary_key#L26)):*
You can do something like `SELECT length(_rowid) FROM t` to verify that the `_rowid` column exists. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by mjibson** [comment]
Review status: 0 of 2 files reviewed at latest revision, 5 unresolved discussions.

---

*[sql/create.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z09bhdgQ-U0oH2jhv:-K4Z5hDXIWCgxy_dh9mk:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L132)):*
Done.

---

*[sql/create.go, line 134 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z14FVNeZMMqbkETO1:-K4Z6qc8HJCKyV61eVf-:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L134)):*
Done.

---

*[sql/testdata/no_primary_key, line 26 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z1r3RhUy45_gL0NMm:-K4Z60Lrv3sWKYWN1CHN:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/testdata/no_primary_key#L26)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

File: sql/create.go [hunk 2]
```diff
@@ -126,7 +126,28 @@ func (p *planner) CreateTable(n *parser.CreateTable) (planNode, error) {
...
+		// Ensure a Primary Key exists.
+		s := "experimental_unique_int()"
```
Comment (tschottdorf) [inline sql/create.go]: nit: `const s = "experimental_unique_int()"`

Comment (tschottdorf) [inline sql/create.go]: ah, taking the address. That probably ruins it.

**Comment by tamird** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/create.go, line 132 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4YwzflK7TX1jPbEoSw-r2-132:-K4ZC0MsHUo2b03pDaXB:1961398485) ([raw file](https://github.com/cockroachdb/cockroach/blob/38a3dc73b1131c6acde6b715b66abf7a78f6c057/sql/create.go#L132)):*
We should extract a constant, so that when this method is no longer "experimental", this doesn't rot.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by petermattis** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/create.go, line 132 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4YwzflK7TX1jPbEoSw-r2-132:-K4ZCODki_n9RZb7pDuy:1766613625) ([raw file](https://github.com/cockroachdb/cockroach/blob/38a3dc73b1131c6acde6b715b66abf7a78f6c057/sql/create.go#L132)):*
The test would fail if we removed the `experimental_` prefix from the builtin without fixing this code. I fail to see how rot would occur.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by petermattis** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/testdata/no_primary_key, line 10 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z22xbr8ikoHdTw-Ml:-K4ZCf454QJHX5-b1SRc:-1618943912) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/testdata/no_primary_key#L10)):*
I'm fine with addressing this in a future PR, but it would good to add a test case demonstrating the problem to this PR.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by tamird** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/create.go, line 132 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4YwzflK7TX1jPbEoSw-r2-132:-K4ZD-y-wyIEvbGUksol:1669290035) ([raw file](https://github.com/cockroachdb/cockroach/blob/38a3dc73b1131c6acde6b715b66abf7a78f6c057/sql/create.go#L132)):*
"Rot" was the wrong word.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by mjibson** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/create.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4YwzflK7TX1jPbEoSw-r1-130:-K4ZEKBoXdG33vo3qL8j:1582343467) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L130)):*
https://github.com/cockroachdb/cockroach/blob/28a6c079e53c13b11b8c26316efb61d610d4f2e0/sql/structured.go#L246

Is where desc.PrimaryIndex.ColumnIDs is appended inside of the AllocateIDs func. The create function could go through and manually examine all indicies to see if any are primary, but it may be better to keep that logic in one place in AllocateIDs.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by mjibson** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/testdata/no_primary_key, line 10 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z22xbr8ikoHdTw-Ml:-K4ZEsNeq6XLMC82uoxm:-456045590) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/testdata/no_primary_key#L10)):*
Where should I make this test case? If I make it in this PR, it'll have to be commented out with a comment explaining why. What about an open issue/PR with the breaking test case, that we can merge with the fix?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by mjibson** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/create.go, line 132 \[r2\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4YwzflK7TX1jPbEoSw-r2-132:-K4ZFHpUvJ343mqHmuGx:453509156) ([raw file](https://github.com/cockroachdb/cockroach/blob/38a3dc73b1131c6acde6b715b66abf7a78f6c057/sql/create.go#L132)):*
The address taking is indeed why you can't use a const here. I have no opinion on moving the string to a common place so it doesn't break during rename. I'm inclined to leave it as is.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by petermattis** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/testdata/no_primary_key, line 10 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z22xbr8ikoHdTw-Ml:-K4ZFvZPWoDBKMdxaRtr:-2119533699) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/testdata/no_primary_key#L10)):*
I'd add the test case to the end of this file. You don't have to comment it out, just expect the error (`statement error ...`) instead of success and add a `TODO(mjibson): this should succeed` comment. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by tamird** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/create.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4YwzflK7TX1jPbEoSw-r1-130:-K4ZHov5_RD-9FpJd9Wh:1676918102) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L130)):*
I'm not sure what that has to do with anything; why does it prevent you from checking `ColumnNames`?

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by mjibson** [comment]
Review status: 0 of 2 files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/testdata/no_primary_key, line 10 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z22xbr8ikoHdTw-Ml:-K4ZIENnEmFWJeNeWzsZ:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/testdata/no_primary_key#L10)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by tschottdorf** [comment]
Reviewed 1 of 2 files at r1, 1 of 2 files at r2, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions.

---

*[sql/create.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z09bhdgQ-U0oH2jhv:-K4ZHXu9iDYkd3rMrnpH:1062770874) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L132)):*
hm, I thought we added this mostly for the sql tests and now it seems like we worry about performance in a way that would rather have us speed up ID generation ever so slightly, but instead have write hotspots for all such tables. I'm not invested in this, just curious why that's a good choice.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by petermattis** [comment]
LGTM

---

Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/create.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z09bhdgQ-U0oH2jhv:-K4ZIsLcN9YqUHftoJ13:-2137156309) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L132)):*
I'm not sure what the right choice is here. Certainly, the user shouldn't expect any ordering of the rows if they didn't specify a primary key. ID generation speed seems of small importance. The size of the ID field (16-bytes for the UUID vs 8-bytes for an INT) seems more important, but still relatively minor.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-:-K4ZIoHXI5cR5Bushixp:676834233)*

**Comment by mjibson** [comment]
Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/create.go, line 130 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4YwzflK7TX1jPbEoSw-r1-130:-K4ZQpCNwnBzlc_5aUVs:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L130)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by tschottdorf** [comment]
Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/create.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z09bhdgQ-U0oH2jhv:-K4ZUHMcFosO1NryPsqQ:-864651621) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L132)):*
Suppose we could always have a version of `experimental_unique_int()` which reverses the bit order so that it should be distributed enough if we ever have high-performance users of this.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by tschottdorf** [comment]
Reviewed 1 of 1 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/testdata/no_primary_key, line 38 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4ZWgkSfSzX1pFdw7Rx:-K4ZWgkSfSzX1pFdw7Ry:2024291282) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b068f75fe50f63d5cd19de70883bf7b126c8f19/sql/testdata/no_primary_key#L38)):*
Not a big deal, but you could make the `INSERT` fail by giving column `c` the type `STRING`.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by petermattis** [comment]
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[sql/create.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z09bhdgQ-U0oH2jhv:-K4ZWnd6_YVuZRpuuQio:1013635882) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L132)):*
Yeah. Might be worth having an `internal_rowid()` function. 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by mjibson** [comment]
Review status: 1 of 2 files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[sql/create.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z09bhdgQ-U0oH2jhv:-K4aUJ5QrjC3PVw9Rb5U:-618359842) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L132)):*
I don't understand the problem with `experimental_unique_int` and why reversing the bit order matters.

---

*[sql/testdata/no_primary_key, line 38 \[r4\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4ZWgkSfSzX1pFdw7Rx:-K4aVF3OqJ810Qsqcu_6:-499146009) ([raw file](https://github.com/cockroachdb/cockroach/blob/0b068f75fe50f63d5cd19de70883bf7b126c8f19/sql/testdata/no_primary_key#L38)):*
Done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by petermattis** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/create.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z09bhdgQ-U0oH2jhv:-K4c0eGMtSl5HriWfl_I:1346060997) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L132)):*
`experimental_unique_int` is composed of a timestamp+node-id. The timestamp occupies the high bits and thus causes the generated IDs to be mostly sequential (modulo clock skew between different nodes). Sequential IDs create a hotspot since insertion of rows into a table will all be directed at the same range. If we reversed the bits the writes would be scattered and could take advantage of more write bandwidth. On the downside, reversing the bits would enforce writes being scattered so the insertion of 2 rows would necessarily hit 2 ranges. I'm not sure if there is a best answer here.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by mjibson** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/create.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z09bhdgQ-U0oH2jhv:-K4cpym26SBFptPfdROn:1642673422) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L132)):*
Makes sense. Are there objections to merging this PR, and we can choose to change the implementation later? I don't care one way or the other, but if we don't know which is better, maybe it would work to merge this and see how things go, then change it if it's bad.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by petermattis** [comment]
Review status: 1 of 2 files reviewed at latest revision, 1 unresolved discussion.

---

*[sql/create.go, line 132 \[r1\]](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286#-K4Z09bhdgQ-U0oH2jhv:-K4d8twswHafueywOGES:2125311391) ([raw file](https://github.com/cockroachdb/cockroach/blob/37990c1244c71f16c1a64c1824534c6491740a97/sql/create.go#L132)):*
I have no objection to merging as-is.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Comment by tamird** [comment]
Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/cockroachdb/cockroach/3286)*

**Pull request merged**